ads/google/core-shared.js

```diff
@@ -0,0 +1,520 @@
+};
+
+const PUB_CONTROL_LAYOUT_PREFIX = 'pub_control_';
+
+const PUB_CONTROL_EXAMPLE = '\n ' +
+  'data-matched-content-rows-num=\"4,2\"\n' +
+  'data-matched-content-columns-num=\"1,6\"\n' +
+  'data-matched-content-ui-type=\"image_stacked,image_card_sidebyside\"';
+
+/**
+ * Configuration of content recommendation unit for current slot. This is the
+ * result of running CoRe responsive logic and values from this config
+ * will be used in ad request.
+ * @record
+ */
+class CoReConfig { // eslint-disable-line no-unused-vars
+  /** see comment on class */
+  constructor() {
+    /** @const {number} */
+    this.slotWidth;
+
+    /** @const {number} */
+    this.slotHeight;
+
+    /**
+     * Number of rows to return in matched content unit. Corresponds to
+     * "cr_col" url param.
+     * @const {number}
+     */
+    this.numberOfRows;
+
+    /**
+     * Number of columns to return in matched content unit. Corresponds to
+     * "cr_row" url param.
+     * @const {number}
+     */
+    this.numberOfColumns;
+
+    /**
+     * Layout type to use for currect matched content slot. Corresponds to
+     * "crui" url param.
+     * @const {!LayoutType}
+     */
+    this.layoutType;
+
+    /**
+     * If not null then it contains an error that some of the provided
+     * parameters are incorrect. The error is intended to be displayed to
+     * developers who setup ad tag. For example it can displayed in console
+     * or thrown as js error. If validation is set other params should be
+     * ignored.
+     * @const {string|undefined}
+     */
+    this.validationError;
+  }
+}
+
+/**
+ * @param {number} availableWidth
+ * @param {boolean} isMobile
+ * @return {!CoReConfig}
+ */
+export function getAutoConfig(availableWidth, isMobile) {
+  if (availableWidth < MIN_PUB_CONTROL_WIDTH_OF_DESKTOP) {
+    if (isMobile) {
+      const layoutType = LayoutType.MOBILE_BANNER_IMAGE_SIDEBYSIDE;
+      const numColumns = 1;
+      const numRows = 12;
+      const slotSize = getLargerAdOneColumnSidebysideSize(
+          availableWidth, layoutType, numColumns, numRows);
+      return {
+        slotWidth: slotSize.width,
+        slotHeight: slotSize.height,
+        numberOfColumns: numColumns,
+        numberOfRows: numRows,
+        layoutType,
+      };
+    } else {
+      const slotSize = getAutoSlotSize(availableWidth);
+      return {
+        slotWidth: slotSize.width,
+        slotHeight: slotSize.height,
+        numberOfColumns: 1,
+        numberOfRows: 13,
+        layoutType: LayoutType.IMAGE_SIDEBYSIDE,
+      };
+    }
+  } else {
+    const slotSize = getAutoSlotSize(availableWidth);
+    return {
+      slotWidth: slotSize.width,
+      slotHeight: slotSize.height,
+      numberOfColumns: 4,
+      numberOfRows: 2,
+      layoutType: LayoutType.IMAGE_STACKED,
+    };
+  }
+}
+
+/**
+ * Get CoRe Pub Control UI Sizes.
+ * @param {number} availableWidth
+ * @param {{
+ *   numberOfRows: (string|undefined),
```

Comment:
You repeat this type multiple times, can you make it a proper typedef?

Reply:
Done. Introduced RawPublisherControlParams typedef.